### PR TITLE
Superheroes secrets setup

### PR DESCRIPTION
**Pull request by Vidal322**
Added a google secret for each superhero with id: superhero-0x-0x-secret.

Gave [iam permissions](https://cloud.google.com/secret-manager/docs/access-control) to each superhero account for its own secret:
    - secretAccessor 
    - secretVersionManager

Due to [Pricing](https://cloud.google.com/secret-manager/pricing) reasons we might need to create a cloud trigger, to disable/destroy older versions of each secret when new ones are created.

Injecting secrets on superhero docker images building in the pipeline

### CODE REVIEW

**Comment by MRita443** [comment]
Could you include a secret for Jarvis too? It'll need one as well!

**Comment by Vidal322** [comment]
@cosmicblackcat that sounds like a good simpler option, however doesn't that allow some superhero to exceed the 6 version per secret free limit before the the cloudbuild.yaml runs?

**Comment by Vidal322** [comment]
Sure @MRita443. Which accounts should have access to it?

**Comment by MRita443** [comment]
> Sure @MRita443. Which accounts should have access to it? 
> 

Just the Jarvis service account please!

**Comment by Vidal322** [comment]
@cosmicblackcat I couldn't find any way to limit the number of versions per secret, I could create a null_service to clean older versions up, however it would only run on the cloudbuild. Perhaps I'm missing something. Any ideas?

File: infrastructure/terraform/main.tf [hunk 1]
```diff
@@ -607,4 +607,136 @@ resource "google_project_iam_member" "strange_firestore_permissions" {
...
+        --filter="state=ENABLED" \
+        --format="value(name)" | tail -n +2 | xargs -I {} gcloud secrets versions destroy {} \
+        --secret=strange-secret
+    EOT
+  }
+
+  triggers = {
+    always_run = "${timestamp()}"
+  }
 }
```
[the remark this answers is not in the document]
Comment (Vidal322) [reply infrastructure/terraform/main.tf:742]: Makes sense, will change.

**Comment by MRita443** [comment]
Actually Jarvis will need two secrets, and this will probably be the case for every superhero: one service account key to read from Nexus, and (in the case of superheroes) an LLM API key (fact check pls: @cosmicblackcat ). Could we switch to that?

**Comment by MRita443** [comment]
> > Actually Jarvis will need two secrets, and this will probably be the case for every superhero: one service account key to read from Nexus, and (in the case of superheroes) an LLM API key (fact check pls: @cosmicblackcat ). Could we switch to that?
> 
> This makes sense @MRita443 . Maybe create 5 per superhero, just so that there's margin if additional secrets are needed @Vidal322? Because there's also merger and possibly other services... Or this could be done on-demand, with students modifying the terraform file as they need. But I'm expecting a lot of delays and pipeline breakages if we go this way... Should probably just be a foreach loop in the terraform file, simple enough I believe.

I think also because of the pricing concerns we should try to find an intermediate, maybe 3 secrets to be one more than the mandatory ones? And then if more than that which I assume is unlikely, then we can open a PR

**Comment by Vidal322** [comment]
@cosmicblackcat , now that I look at it, isn't this implementation going to exceed the free limits?
```
8 classes * 4 groups * 1 secret each * 1 version = 32.
```
image

**Comment by Vidal322** [comment]
> Also, is that data per service account or per project? If it is per service account, that should not be an issue.

It's per billing account, I'm assuming there's only one for the whole project?

**Comment by Vidal322** [comment]
> You have a point. Do you have any ideas on how to mitigate this? Maybe store not-so-critical info in Vault instead? Just an idea, and definitely not an ideal solution, but I would like to hear from you.

Don't really know what the ups and downs are for each of the alternatives, would having the secrets stored in a cloud bucket with  manually handled encryption be too big of a stretch? 
Don't really know what the difference between those 2 options are, and/or what would be the hardest to implement.
I can search other secret manager alternatives but I don't think I can find free plans with 30+ secret versions available.

**Review by JaimeFRF — APPROVED** [review]

**Pull request merged**